### PR TITLE
[DNM] *: support read statistic upload to pd

### PR DESCRIPTION
**Pull request by choleraehyq**
Signed-off-by: Cholerae Hu <huyingqian@pingcap.com>

### CODE REVIEW

File: src/storage/mvcc/reader.rs [hunk 1]
```diff
@@ -222,7 +224,6 @@ impl<'a> MvccReader<'a> {
             match try!(self.seek_write(key, ts)) {
                 Some((commit_ts, mut write)) => match write.write_type {
                     WriteType::Put => {
-                        self.statistics.write.processed += 1;
```
Comment (siddontang) [inline src/storage/mvcc/reader.rs]: why remove it?

Comment (choleraehyq) [inline src/storage/mvcc/reader.rs]: Duplicate. Write processed has already be added in seek_write.

File: src/coprocessor/endpoint.rs [hunk 2]
```diff
@@ -67,35 +68,55 @@ pub struct Host {
...
+    low_priority_pool: ThreadPool<CopContext>,
+    high_priority_pool: ThreadPool<CopContext>,
     max_running_task_count: usize,
 }
 
-// TODO: remove these dummy structures.
 #[derive(Clone)]
-struct DummyContext {}
+struct CopContext {
+    raftstore_sender: Option<Sender<Statistics>>,
```
Comment (siddontang) [inline src/coprocessor/endpoint.rs]: raftstore should not know the storage Statistics

File: src/coprocessor/endpoint.rs [hunk 3]
```diff
@@ -67,35 +68,55 @@ pub struct Host {
...
-unsafe impl Send for DummyContext {}
+unsafe impl Send for CopContext {}
 
-impl Context for DummyContext {
+impl Context for CopContext {
     fn on_task_started(&mut self) {}
     fn on_task_finished(&mut self) {}
-    fn on_tick(&mut self) {}
+    fn on_tick(&mut self) {
+        if let Some(ref sender) = self.raftstore_sender {
```
Comment (siddontang) [inline src/coprocessor/endpoint.rs]: the default timeout is  20ms. If we do nothing, we still need to send the metrics.

I think we should enlarge the default timeout for the thread pool, maybe 500ms or 1s.

File: src/coprocessor/endpoint.rs (hunk 2 above)
Comment (choleraehyq) [inline src/coprocessor/endpoint.rs]: Heartbeat was send to pd from raftstore.

Comment (siddontang) [inline src/coprocessor/endpoint.rs]: raftstore only needs read statistics, but not the whole Statistics. 

Btw, how do we know the read statistics for different region?

File: src/coprocessor/endpoint.rs [hunk 4]
```diff
@@ -69,13 +72,22 @@ pub struct Host {
...
+
+#[derive(Clone)]
 struct CopContext {
-    task_count: u64,
+    task_count: i64,
+    raftstore_sender: Option<Sender<ReadStats>>,
     select_stats: Statistics,
     index_stats: Statistics,
     dag_stats: Statistics,
+    // region_id => read statistics of this region.
```
Comment (zhangjinpeng1987) [inline src/coprocessor/endpoint.rs]: Remove this line.

File: src/coprocessor/endpoint.rs [hunk 5]
```diff
@@ -197,9 +232,13 @@ impl Host {
                 CommandPri::Normal => &mut self.pool,
             };
             pool.execute(move |ctx: &mut CopContext| {
+                let region_id = req.req.get_context().get_region_id();
                 let stats = end_point.handle_request(req);
                 ctx.task_count += 1;
                 ctx.add_statistics(type_str, &stats);
+                let m = Arc::make_mut(&mut ctx.read_stats);
```
Comment (zhangjinpeng1987) [inline src/coprocessor/endpoint.rs]: Can we use `let e = ctx.read_stats.entry(region_id).or_insert((0,0))` ?

File: src/server/node.rs [hunk 6]
```diff
@@ -316,6 +319,7 @@ where
         Err(box_err!("check cluster bootstrapped failed"))
     }
 
+    #[allow(too_many_arguments)]
```
Comment (zhangjinpeng1987) [inline src/server/node.rs]: Not allowed. :)

Comment (choleraehyq) [reply src/server/node.rs]: https://github.com/pingcap/tikv/blob/ba174ad54c8e2ff9d45979a4f6a65ca01901308a/src/coprocessor/codec/mysql/time.rs#L43

Refactor this function to reduce arguments will make this signature less readable.

File: src/coprocessor/endpoint.rs [hunk 7]
```diff
@@ -69,13 +72,20 @@ pub struct Host {
     max_running_task_count: usize,
 }
 
+pub type ReadStats = Arc<HashMap<u64, (u64, u64)>>;
```
Comment (siddontang) [inline src/coprocessor/endpoint.rs]: why using Arc here?

File: src/bin/tikv-server.rs [hunk 8]
```diff
@@ -206,6 +206,9 @@ fn run_raft_server(pd_client: RpcClient, cfg: &TiKvConfig) {
             raft_db_cf_opts,
         ).unwrap_or_else(|s| fatal!("failed to create raft engine: {:?}", s)),
     );
+
+    let (read_stat_sender, read_stat_recv) = mpsc::channel();
```
Comment (siddontang) [inline src/bin/tikv-server.rs]: why not using raftstore channel directly?

File: src/raftstore/store/store.rs [hunk 9]
```diff
@@ -1860,14 +1866,34 @@ impl<T: Transport, C: PdClient> Store<T, C> {
     }
 
     fn on_pd_heartbeat_tick(&mut self, event_loop: &mut EventLoop<Self>) {
+        let mut read_statistic: HashMap<u64, (u64, u64)> = HashMap::default();
+        if let Some(ref recv) = self.read_stat_recv {
+            while let Ok(m) = recv.try_recv() {
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: bad to handle all read here.

File: src/coprocessor/endpoint.rs (hunk 7 above)
Comment (siddontang) [reply src/coprocessor/endpoint.rs]: please define a structure, not a tuple.

**Comment by nolouch** [comment]
I will post a new PR for this later.

**Pull request closed without merging**